### PR TITLE
solve the problem of duplicate comments and blank lines 

### LINKED ISSUE #5000: Text配置修改，含有注释的情况下，撤销修改导致后续存在自动添加重复注释
- [x] 我已经检查过[discussions](https://github.com/ctripcorp/apollo/discussions)
- [x] 我已经搜索过[issues](https://github.com/ctripcorp/apollo/issues)
- [x] 我已经仔细检查过[FAQ](https://www.apolloconfig.com/#/zh/faq/common-issues-in-deployment-and-development-phase)

**描述bug**
Text配置修改，含有注释的情况下，撤销修改导致后续存在自动添加重复注释

**复现**

通过如下步骤可以复现：

1. 通过文本编辑写配置和注释并发布
```
#注释第一行
key2 = value2
#注释第二行
key3 = value3
```
image

2. 文本编辑移除key3配置，提交但不发布
image

3. 点击撤销按钮
image
image
4. 文本编辑追加配置提交
image

6. 再次文本编辑发现有自动添加
image

**期望**

简介明了地描述你希望正常情况下应该发生什么
不应该出现文本编辑自动添加一些注释的情况

版本：2.0.1

### PR DESCRIPTION
**Pull request by youngzil**
1. solve the problem of duplicate comments and blank lines
2. reduce the data growth of the item table
3. in the same case, the size of ItemChangeSets can be reduced
4. when revoke configuration, if there is only one comment or blank line, it will not be deleted.

## What's the purpose of this PR

XXXXX

## Which issue(s) this PR fixes:
Fixes #5000

## Brief changelog

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [ ] Read the [Contributing Guide](https://github.com/apolloconfig/apollo/blob/master/CONTRIBUTING.md) before making this pull request.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit tests to verify the code.
- [ ] Run `mvn clean test` to make sure this pull request doesn't break anything.
- [ ] Update the [`CHANGES` log](https://github.com/apolloconfig/apollo/blob/master/CHANGES.md).

## Summary by CodeRabbit

- **New Features**
	- Enhanced handling of comment and blank items for improved processing efficiency.
	- Streamlined logic for managing deleted items, ensuring only valid entries are processed.
	- Added functionality to address duplicate comments and blank lines in the Apollo system.

- **Bug Fixes**
	- Resolved issues related to duplicate comments and blank lines in configuration management.

- **Refactor**
	- Simplified code structure for better maintainability and clarity in various methods related to item management.

### CODE REVIEW

**Comment by youngzil** [comment]
// TODO 
code：`com.ctrip.framework.apollo.portal.service.ItemService#revokeItem`
Remaining issues：
When revoking a configuration, the number of configuration lines will be reset, which will cause confusion between the number of comments and configuration lines. This cannot be solved at present.

**Comment by youngzil** [comment]
**### Add a question about revoke configuration**

1. The initial configuration is as follows

image

2. The second step only deletes the fourth line, and the comment on the third line is still retained.
3. But when I revoke the configuration, I find that the comment on the third line is also deleted.

image

4. The reasons why this happens are as follows：
L218：oldKeyMapItem contains a comment or a blank line
L241：The comment or blank line in oldKeyMapItem will be deleted.

So  I removed this comment or blank line in #5232 

  https://github.com/apolloconfig/apollo/blob/31e6486791de3ad60f059f34957b343839aa5ebb/apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/ItemService.java#L218C1-L219C1
https://github.com/apolloconfig/apollo/blob/31e6486791de3ad60f059f34957b343839aa5ebb/apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/ItemService.java#L241

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/txtresolver/PropertyResolver.java [hunk 1]
```diff
@@ -122,16 +144,18 @@ private String[] parseKeyValueFromItem(String item) {
   }
 
   private void handleCommentLine(Long namespaceId, ItemDTO oldItemByLine, String newItem, int lineCounter, ItemChangeSets changeSets) {
-    String oldComment = oldItemByLine == null ? "" : oldItemByLine.getComment();
-    //create comment. implement update comment by delete old comment and create new comment
-    if (!(isCommentItem(oldItemByLine) && newItem.equals(oldComment))) {
+    if(null == oldItemByLine ){
       changeSets.addCreateItem(buildCommentItem(0L, namespaceId, newItem, lineCounter));
+    }else if(!StringUtils.equals(oldItemByLine.getComment(), newItem) || lineCounter != oldItemByLine.getLineNum()) {
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/txtresolver/PropertyResolver.java:149]:
```suggestion
    } else if(!StringUtils.equals(oldItemByLine.getComment(), newItem) || lineCounter != oldItemByLine.getLineNum()) {
```

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/txtresolver/PropertyResolver.java [hunk 2]
```diff
@@ -122,16 +144,18 @@ private String[] parseKeyValueFromItem(String item) {
   }
 
   private void handleCommentLine(Long namespaceId, ItemDTO oldItemByLine, String newItem, int lineCounter, ItemChangeSets changeSets) {
-    String oldComment = oldItemByLine == null ? "" : oldItemByLine.getComment();
-    //create comment. implement update comment by delete old comment and create new comment
-    if (!(isCommentItem(oldItemByLine) && newItem.equals(oldComment))) {
+    if(null == oldItemByLine ){
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/txtresolver/PropertyResolver.java:147]:
```suggestion
    if (null == oldItemByLine){
```

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/txtresolver/PropertyResolver.java [hunk 3]
```diff
@@ -122,16 +144,18 @@ private String[] parseKeyValueFromItem(String item) {
...
       changeSets.addCreateItem(buildCommentItem(0L, namespaceId, newItem, lineCounter));
+    }else if(!StringUtils.equals(oldItemByLine.getComment(), newItem) || lineCounter != oldItemByLine.getLineNum()) {
+      changeSets.addUpdateItem(buildCommentItem(oldItemByLine.getId(), namespaceId, newItem, lineCounter));
     }
   }
 
   private void handleBlankLine(Long namespaceId, ItemDTO oldItem, int lineCounter, ItemChangeSets changeSets) {
-    if (!isBlankItem(oldItem)) {
+    if(null == oldItem ){
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/txtresolver/PropertyResolver.java:155]:
```suggestion
    if (null == oldItem ){
```

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/txtresolver/PropertyResolver.java [hunk 4]
```diff
@@ -122,16 +144,18 @@ private String[] parseKeyValueFromItem(String item) {
...
       changeSets.addCreateItem(buildBlankItem(0L, namespaceId, lineCounter));
+    }else if (lineCounter != oldItem.getLineNum()) {
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/component/txtresolver/PropertyResolver.java:157]:
```suggestion
    } else if (lineCounter != oldItem.getLineNum()) {
```

**Review by nobodyiam — COMMENTED** [review]
1. Please update the CHANGES.md
2. I tested the new patch but find the line num is wrong after revoking the configurations
the original configuration
image
after line 4 is deleted
image
after the configuration is revoked
image

**Comment by youngzil** [comment]
> // TODO code：`com.ctrip.framework.apollo.portal.service.ItemService#revokeItem` Remaining issues： When revoking a configuration, the number of configuration lines will be reset, which will cause confusion between the number of comments and configuration lines. This cannot be solved at present.


I mentioned this problem in my reply above, but I don't have a good solution to solve it completely.

I submitted some small optimizations for this problem. With a small number of changes, the line num can be kept correct as much as possible.
But this code cannot solve the situation where a lot of comments and blank lines are added before undoing.
It also cannot solve the situation where the number of lines of comments changes a lot.
So this is not a complete solution

The code is as follows：
https://github.com/youngzil/apollo/blob/859bbdbf00934e6f308c8685bee616c106b42ebe/apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/ItemService.java#L233C6-L238C8

**Review by nobodyiam — APPROVED** [review]
LGTM

**Pull request merged**